crypto/evp/digest.c

```diff
@@ -617,23 +617,37 @@ int EVP_MD_CTX_copy_ex(EVP_MD_CTX *out, const EVP_MD_CTX *in)
         return 0;
     }
 
-    evp_md_ctx_reset_ex(out, 1);
-    digest_change = (out->fetched_digest != in->fetched_digest);
-    if (digest_change && out->fetched_digest != NULL)
-        EVP_MD_free(out->fetched_digest);
-    *out = *in;
-    /* NULL out pointers in case of error */
-    out->pctx = NULL;
-    out->algctx = NULL;
+    if (in->algctx != NULL && out->algctx != NULL && in->digest->copyctx != NULL &&
+        out->digest == in->digest && out->fetched_digest == in->fetched_digest && out->reqdigest == in->reqdigest) {
```

Comment:
This could be much weaker condition - `out->digest == in->digest && in->digest->copyctx != NULL` as all the others follow from that or aren't relevant. We are here in the non-legacy branch of the code and in->digest cannot be NULL as that was also checked already above. We could even check just whether `out->digest != NULL && out->digest->prov == in->digest->prov && in->digest->copyctx != NULL` as if the provider is the same the digest can be copied too.